### ISSUE TITLE
[Question] Listening for Eia608/Cea608 Closed Captions using TextRenderer

### ISSUE THREAD

**Issue by rileymacdonald** [google/ExoPlayer#2620]
I'm upgrading from ExoPlayer `r.1.5.9` to `r2.2.0`. Previously in ExoPlayer `r1.5.9` I listened for `CEA-608` / `EIA-608` captions using `Eia608TrackRenderer`:
```java
new Eia608TrackRenderer(sampleSource, new Consumer<List<Cue>>() {
    @Override
    public void accept(List<Cue> cues) {
        // Handle captions
        final List<CharSequence> lines = new ArrayList<>(cues.size());
        for (final Cue cue : cues) lines.add(cue.text);
        observer.captionsChanged(lines);
    }
}
```
I've been unable to determine how to listen for these captions using ExoPlayer2. Here's the ExoPlayer2 instance I'm using:
```java
ExoPlayerFactory.newInstance(
    renderers, 
    new DefaultTrackSelector(new AdaptiveVideoTrackSelection.Factory(new DefaultBandwidthMeter())),
    new DefaultLoadControl(new DefaultAllocator(true, bufferChunkSize)));
```
The `renderers` (above) are `new MediaCodecVideoRenderer(appContext, MediaCodecSelector.DEFAULT)`, `new MediaCodecAudioRenderer(MediaCodecSelector.DEFAULT)`, `new MetadataRenderer(...)`, and finally:
```java
new TextRenderer(new TextRenderer.Output() {
    @Override public void onCues(List<Cue> cues) {
        final List<CharSequence> lines = new ArrayList<>(cues.size());
        for (final Cue cue : cues) lines.add(cue.text);
        observer.captionsChanged(lines);
    }
}, outputLooper);
```
I observe `WebVTT` captions from the `TextRenderer.Output` while playing back the Apple sample stream: https://devimages.apple.com.edgekey.net/streaming/examples/bipbop_16x9/bipbop_16x9_variant.m3u8.

While playing back the media (`new HlsMediaSource(...)`) in question `MetadataRenderer.Output` is working but I'm not receiving any `TextRenderer.Output`. Do I need to provide something such as a subtitle decoder? Are the default instances I'm using above adequate? Is there any additional documentation regarding the TextRenderer I can refer to? 

**I assume this is not an issue with the media in question because it's working fine in ExoPlayer r1.5.9. Something with my implementation must be wrong.**.

**Comment by ojw28** [comment]
You probably just need to ensure the track is selected. A caption track will only be selected by default if the media indicates that this should be the case. Else you have to select it yourself. Note that in the Apple stream you reference one of the caption tracks is marked `DEFAULT=YES`, which is why that one is enabled by default.

If you take a look at the ExoPlayer demo app, you should see that you're able to select from the available text tracks by clicking the "Text" button. For the Apple sample stream you should see both VTT and 608 tracks, and be able to select between them. Once a 608 track is selected you'll receive the output. So I suggest you take a look at how the demo app does that, and also look at `DefaultTrackSelector`.

**Comment by rileymacdonald** [comment]
Thanks! You're correct, I used a `MappingTrackSelector.SelectionOverride` along with `DefaultTrackSelector.setSelectionOverride()` to force selecting the desired text track. The `TextRenderer.Output` is now working as expected. **This issue can be closed.**

Although this issue can be closed I had to implement a small workaround due to a `NoSuchMethodException` in `SubtitleDecoderFactory` that I feel I should mention.

Here's the `Format` argument handed to the `createDecoder()` method:
```
Format(null, null, application/cea-608, -1, null, [-1, -1, -1.0], [-1, -1])
```

The full stacktrace:
```
03-30 15:53:01.296 24721-25368/com.vmn.android.sample E/ExoPlayerImplInternal: Internal runtime error.
                                                                               java.lang.IllegalStateException: Unexpected error instantiating decoder
                                                                                   at com.google.android.exoplayer2.text.SubtitleDecoderFactory$1.createDecoder(SubtitleDecoderFactory.java:90)
                                                                                   at com.google.android.exoplayer2.text.TextRenderer.onStreamChanged(TextRenderer.java:112)
                                                                                   at com.google.android.exoplayer2.BaseRenderer.replaceStream(BaseRenderer.java:99)
                                                                                   at com.google.android.exoplayer2.BaseRenderer.enable(BaseRenderer.java:81)
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.enableRenderers(ExoPlayerImplInternal.java:1371)
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.reselectTracksInternal(ExoPlayerImplInternal.java:790)
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:328)
                                                                                   at android.os.Handler.dispatchMessage(Handler.java:98)
                                                                                   at android.os.Looper.loop(Looper.java:148)
                                                                                   at android.os.HandlerThread.run(HandlerThread.java:61)
                                                                                   at com.google.android.exoplayer2.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40)
                                                                                Caused by: java.lang.NoSuchMethodException: <init> [class java.lang.String, int]
                                                                                   at java.lang.Class.getConstructor(Class.java:528)
                                                                                   at java.lang.Class.getConstructor(Class.java:492)
                                                                                   at com.google.android.exoplayer2.text.SubtitleDecoderFactory$1.createDecoder(SubtitleDecoderFactory.java:81)
                                                                                   at com.google.android.exoplayer2.text.TextRenderer.onStreamChanged(TextRenderer.java:112) 
                                                                                   at com.google.android.exoplayer2.BaseRenderer.replaceStream(BaseRenderer.java:99) 
                                                                                   at com.google.android.exoplayer2.BaseRenderer.enable(BaseRenderer.java:81) 
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.enableRenderers(ExoPlayerImplInternal.java:1371) 
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.reselectTracksInternal(ExoPlayerImplInternal.java:790) 
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:328) 
                                                                                   at android.os.Handler.dispatchMessage(Handler.java:98) 
                                                                                   at android.os.Looper.loop(Looper.java:148) 
                                                                                   at android.os.HandlerThread.run(HandlerThread.java:61) 
                                                                                   at com.google.android.exoplayer2.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40) 
```

Since I'm only looking to support Cea608 captions at this time I wrote my own implementation of `SubtitleDecoderFactory` to provide the `Cea608Decoder` I needed and handed it to my `TextRenderer`.

**Comment by ojw28** [comment]
The `NoSuchMethodException` issue occurs when proguard removes the constructor being invoked. It should have been fixed by https://github.com/google/ExoPlayer/commit/1eede5d5bb9c0271592b05dbd0db320214d09b18, but that fix was present in 2.2.0. Is that definitely the version you're using? If so, does your application's proguard configuration somehow override the default one that we're specifying for the library? If so, that's likely the issue.

**Comment by rileymacdonald** [comment]
Thanks! My application was responsible for removing this method. When building ExoPlayer I'm producing both `aar` and `jar` artifacts. Gradle was consuming the `jar` by default which doesn't appear to include proguard information. I specified `aar` using `@aar` on the ExoPlayer `compile` dependency and the methods were no longer removed by my application proguard configuration.
